pkg/operators/hub/controller.go

```diff
@@ -2,34 +2,295 @@ package hub
+		nucleusClient:                  nucleusClient,
+		nucleusLister:                  nucleusInformer.Lister(),
+		currentHubDeploymentGeneration: 0,
+	}
+
+	return factory.New().WithSync(controller.sync).
+		ResyncEvery(3*time.Minute).
+		WithInformers(nucleusInformer.Informer()).
+		ToController("NucleusHubController", recorder)
+}
+
+// hubConfig is used to render the template of hub manifests
+type hubConfig struct {
+	HubCoreName       string
+	HubCoreNamespace  string
+	RegistrationImage string
+}
+
+func (n *nucleusHubController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	hubCoreName := controllerContext.QueueKey()
+	klog.V(4).Infof("Reconciling HubCore %q", hubCoreName)
+
+	hubCore, err := n.nucleusLister.Get(hubCoreName)
+	if err != nil {
+		return err
 	}
+	hubCore = hubCore.DeepCopy()
 
-	return factory.New().WithSync(controller.sync).ToController("NucleusHubController", recorder)
+	// Update finalizer at first
+	if hubCore.DeletionTimestamp.IsZero() {
+		hasFinalizer := false
+		for i := range hubCore.Finalizers {
+			if hubCore.Finalizers[i] == nucleusHubFinalizer {
+				hasFinalizer = true
+				break
+			}
+		}
+		if !hasFinalizer {
+			hubCore.Finalizers = append(hubCore.Finalizers, nucleusHubFinalizer)
+			_, err := n.nucleusClient.Update(ctx, hubCore, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	// HubCore is deleting, we remove its related resources on hub
+	if !hubCore.DeletionTimestamp.IsZero() {
+		if err := n.cleanUp(ctx, controllerContext, hubCore.Name, nucluesHubCoreNamespace); err != nil {
+			return err
+		}
+		return n.removeWorkFinalizer(ctx, hubCore)
+	}
+
+	config := hubConfig{
+		HubCoreName:       hubCore.Name,
+		HubCoreNamespace:  nucluesHubCoreNamespace,
+		RegistrationImage: hubCore.Spec.RegistrationImagePullSpec,
+	}
+
+	clientHolder := resourceapply.NewKubeClientHolder(n.kubeClient).WithAPIExtensionsClient(n.apiExtensionClient)
+	// Apply static files
+	resourceResults := resourceapply.ApplyDirectly(
+		clientHolder,
+		controllerContext.Recorder(),
+		func(name string) ([]byte, error) {
+			return assets.MustCreateAssetFromTemplate(name, bindata.MustAsset(filepath.Join("", name)), config).Data, nil
+		},
+		"manifests/hub/0000_00_clusters.open-cluster-management.io_spokeclusters.crd.yaml",
+		"manifests/hub/0000_00_work.open-cluster-management.io_workloads.crd.yaml",
+		"manifests/hub/hub-clusterrole.yaml",
+		"manifests/hub/hub-clusterrolebinding.yaml",
+		"manifests/hub/hub-namespace.yaml",
+		"manifests/hub/hub-serviceaccount.yaml",
+	)
+	errs := []error{}
+	for _, result := range resourceResults {
+		if result.Error != nil {
+			errs = append(errs, fmt.Errorf("%q (%T): %v", result.File, result.Type, result.Error))
+		}
+	}
+
+	// Render deployment manifest and apply
+	err = n.applyDeployment(config, controllerContext)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	conditions := &hubCore.Status.Conditions
+	if len(errs) == 0 {
```

Comment:
followup: use Available, Progressing, Degraded like openshift operators.

You can start with lower level FooAvailable and FooDegraded which make more sense here.